pyat/at/physics/energy_loss.py

```diff
@@ -81,6 +84,59 @@ def tracking(ring):
         raise AtError('Invalid method: {}'.format(method))
 
 
+def get_timelag_fromU0(ring, method=ELossMethod.INTEGRAL, cavpts=None):
+    """
+    Get the TimeLag attribute of RF cavities based on frequency,
+    voltage and energy loss per turn, so that the synchronous phase is zero.
+    An error occurs if all cavities do not have the same frequency.
+    Used in set_cavity_phase()
+
+    PARAMETERS
+        ring        lattice description
+
+    KEYWORDS
+        method=ELossMethod.INTEGRAL
+                            method for energy loss computation.
+                            See "get_energy_loss".
+        cavpts=None         Cavity location. If None, use all cavities.
+                            This allows to ignore harmonic cavities.
+    RETURN
+        timelag
+    """
+    if cavpts is None:
+        cavpts = get_refpts(ring, RFCavity)
+    u0 = get_energy_loss(ring, method=method)
+    try:
+        rfv = ring.get_rf_voltage(cavpts=cavpts)
+        freq = ring.get_rf_frequency(cavpts=cavpts)
+        tl0 = ring.get_rf_timelag(cavpts=cavpts)
+    except AtError:
+        freq = numpy.array([cav.Frequency for cav in ring.select(cavpts)])
+        rfv = numpy.array([cav.Voltage for cav in ring.select(cavpts)])
+        tl0 = numpy.array([cav.TimeLag for cav in ring.select(cavpts)])
+        if u0 > numpy.sum(rfv):
+            raise AtError('Not enough RF voltage: unstable ring')
+        ctmax = 1/numpy.amin(freq)*clight/2
+        tt0 = tl0[numpy.argmin(freq)]
+        eq = lambda x: numpy.sum(rfv*numpy.sin(2*pi*freq*(x+tl0)/clight))-u0
+        deq = lambda x: numpy.sum(rfv*numpy.cos(2*pi*freq*(x+tl0)/clight))
+        zero_diff = least_squares(deq, -tt0+ctmax/2,
+                                  bounds=(-tt0, ctmax-tt0)).x[0]
+        if numpy.sign(deq(zero_diff-1.0e-6)) > 0:
+            ts = least_squares(eq, (zero_diff-tt0)/2,
+                                 bounds=(-tt0, zero_diff)).x[0]
+        else:
+            ts = least_squares(eq, (ctmax-tt0+zero_diff)/2,
+                                 bounds=(zero_diff, ctmax-tt0)).x[0]
+        timelag = ts+tl0
+    else:
+        if u0 > rfv:
+            raise AtError('Not enough RF voltage: unstable ring')
+        timelag = clight/(2*pi*freq)*numpy.arcsin(u0/rfv) * \
```

Comment:
`cavpts` may be a boolean mask or a test function. So `len(cavpts)` is not always the number of refpoints. You can use `refpts_len(ring,refpts)` or `ring.refcount(refpts)` to get the number. Or better, keep it as a scalar, it should be broadcast to the right size when necessary.

Reply:
Changed to get_cells
I don't like that the function can return different "types" depending on the case so I prefer to leave it as is and return always a scalar